### ISSUE TITLE
add spatial softmax helper?

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#626]
It seems like semantic segmentation is becoming a popular task, so there are quite a few questions on the mailing list lately about doing per-pixel classification. This currently requires some trickery because Theano's softmax function expects 2D input, so you have to shuffle and reshape your 'output maps' for this to work (latest discussion: https://groups.google.com/forum/#!topic/lasagne-users/6P1bodBbhWs )

Perhaps we should add a helper function or layer `SpatialSoftmax`, that does this sequence of operations? Or more generally we could have `Softmax1D`, `Softmax2D` etc.?

One thing we have to be careful about is that it is not confused with a softmax across spatial locations (rather than a per-pixel softmax operation), which is sometimes used for attention mechanisms. Perhaps we should provide a helper for that use case as well.

**Comment by f0k** [comment]
Maybe we should have a general `nonlinearities.Softmax` class that accepts a list of axes and performs the necessary reshaping, and then some predefined instances such as the original `softmax` that accommodate the most common use cases.

> One thing we have to be careful about is that it is not confused with a softmax across spatial locations (rather than a per-pixel softmax operation), which is sometimes used for attention mechanisms. Perhaps we should provide a helper for that use case as well.

Good point. We should be careful about the naming then. What about `softmax_per_location()` (e.g., one classification per pixel, or per voxel) and `softmax_over_locations()` (normalizing the softmax over spatial locations)? Should the original `softmax()` stay limited to flattened input and give a meaningful error for too many input dimensions?

**Comment by benanne** [comment]
All good point, clearly this issue can use some more brainstorming so we get it right :)

**Comment by krikru** [comment]
Until there is a working spatial softmax layer, [here is a workaround](https://groups.google.com/d/msg/lasagne-users/OQR19fuPzbM/5p1lF-PJAQAJ) that uses reshaping of the 'output maps'.

**Comment by f0k** [comment]
> What about `softmax_per_location()` (e.g., one classification per pixel, or per voxel) and `softmax_over_locations()` (normalizing the softmax over spatial locations)?

In #710, we agreed on `softmax_per_location()` and `softmax_per_channel()`.

**Comment by ahundt** [comment]
I came across this, and thought you may be interested in this relevant tf issue: https://github.com/tensorflow/tensorflow/issues/6271

**Comment by f0k** [comment]
> I came across this and thought you may be interested in this relevant tf issue

Thanks! So even the TF developers endorse reshaping/transposing, applying the standard softmax and reshaping back. I'm still contemplating whether this should become a Theano feature instead, so it can use more efficient implementations when available (e.g., cuDNN implements spatial softmax with its CUDNN_SOFTMAX_MODE_CHANNEL, which does a softmax over channels per location).

**Comment by christopher-beckham** [comment]
This was opened recently for Theano, which is exciting!:

https://github.com/Theano/Theano/issues/5712

I too am interested in this since I'm working on some segmentation stuff at the moment. For the time being, /selfplug I wrote a spatial softmax layer that uses the CuDNN softmax (that is, if CuDNN is available, as per f0k's suggestion), otherwise it uses the reshape/dimshuffle trick:

https://gist.github.com/christopher-beckham/362caad79a253a7bde3a7a0e44f11775

I ran some tests and it appears to function as intended, but please let me know if there is a bug somewhere.

**Comment by f0k** [comment]
> I ran some tests and it appears to function as intended, but please let me know if there is a bug somewhere.

Looks good, thanks for sharing! The only potential trouble is that you assume `self.input_shape` to be fully defined. You should use the symbolic shape (`input.shape`) for dimensions that are `None` (or just always use the symbolic shape because it's easier, and use the full `input.shape` instead of `(-1,h,w,c)`).

**Comment by christopher-beckham** [comment]
Thanks, I will make those changes.

One thing I completely forgot though... I think I would have to use an intermediate output of that layer to compute the cross-entropy, since cross-entropy expects a 2d tensor for the predictions/targets.

I also stumbled across a 'cudnn.SpatialCrossEntropyCriterion' here (though I don't use Torch nor Lua):

https://github.com/soumith/cudnn.torch/blob/master/SpatialCrossEntropyCriterion.lua

Does this mean CuDNN has its own `categorical_crossentropy` for this use case, or am I mistaken?

**Comment by f0k** [comment]
> I think I would have to use an intermediate output of that layer to compute the cross-entropy, since cross-entropy expects a 2d tensor for the predictions/targets.

I think it'd be cleaner to have a spatial categorical CE loss. (API-wise, having `lasagne.nonlinearities.Softmax` and `lasagne.objectives.categorical_crossentropy` both with an `axes` argument would be the cleanest solution.) Note that this is only required for integer targets, for one-hot targets, the loss doesn't even need to know the axes. Integer targets are a little more efficient because they use a lookup instead of a large number of multiplications with 1 and 0. But I think many frameworks don't care.

> Does this mean CuDNN has its own `categorical_crossentropy` for this use case, or am I mistaken?

Short answer: No.
Long answer: The softmax/crossentropy business is a bit more complicated than it seems. As long as you're dealing with equations, you can just write down the softmax equation and plug it into the cross-entropy equation, and that's how it's modelled in Theano (and Lasagne). When implementing it, you need to be careful: The exp() in the softmax can overflow, so typically you'll deduct the maximal input value from all inputs before exponentiation. The cross-entropy takes the log() again, so if the exp() underflows, you'll get minus infinity -- to avoid this, typically you'll compute the logsoftmax instead (using [logsumexp](https://en.wikipedia.org/wiki/LogSumExp) to compute the denominator, and then subtracting that from each input). Finally, to obtain the gradient of the loss wrt. the softmax input, you don't want to backprop through the cross-entropy, then backprop through the softmax, but directly compute the gradient as y - t (where y is the softmax output, and t is the target).
That's why many other frameworks just combine softmax and loss into a single layer. Theano has optimizers that recognize the combination of a softmax and log to produce a logsoftmax (which is also provided by cuDNN). I haven't checked what happens to the gradients. (For sigmoid and cross-entropy, Theano obtains a gradient expression that is equivalent to y - t, but [more complex than needed](https://groups.google.com/d/msg/theano-dev/jijOYq1S9Ao/-7odVO_nCQAJ), could be the same for softmax.)
To finally come back to the question, cuDNN doesn't have a categorical crossentropy implementation because it's not needed -- frameworks will use cuDNN to compute the logsoftmax and then compute the crossentropy loss and the gradient themselves.

**Comment by f0k** [comment]
> Note that this is only required for integer targets, for one-hot targets, the loss doesn't even need to know the axes.

Rethinking this, it depends on what we want `categorical_crossentropy` to return: a loss per datapoint, or a loss per classification decision. Currently that's the same. A loss per datapoint would require summing across all axes except the batch dimension, which would indeed be independent of what axes the softmax was taken over. A loss per classification decision would require summing only across the axes the softmax was taken over, so the categorical cross-entropy would need to know. And the latter would be more flexible in terms of weighting and maxing, and more reasonable for handling segmentation of differently-sized images. So I think `categorical_crossentropy` should get an `axes` argument.

**Comment by f0k** [comment]
Another question is what to default the `axis` argument of a prospective `Softmax()` class and the `categorical_crossentropy` to. With `axis=1`, it would work correctly for dense layers and convolutional layers trained for segmentation. With `axis=-1`, it would work correctly for dense layers and recurrent layers meant to classify each time step in each sequence. To not pose a trap, I guess it would be best to default `axis=None` and throw an error for input of more than 2 dimensions, forcing users to specify the correct `axis` argument for their application.

**Comment by christopher-beckham** [comment]
Is it just me or does it seem like things could get a bit complicated for integer targets? In the conventional case, the predicted output is a tensor of shape `(bs, k)` (where `bs` = batch size, `k` = number of classes), so the corresponding integer targets are a vector of shape `(bs,)`. If however the predicted output is a 3d tensor `(bs, f, k)` (where `f` is some extra dimension), then the targets are now an 'integer matrix' of size `(bs, f)`, so we'd have to find a nice and clean way to use that matrix to index the tensor. What if we also have the case where the predicted tensor is of shape `(bs, k, f)` (i.e., the softmax is now done over the 2nd axis), or if the predicted tensor was of shape `(bs, k1, k2)` (where the softmax is over both `k1` and `k2`)?

**Comment by f0k** [comment]
> Is it just me or does it seem like things could get a bit complicated for integer targets?

Yes, I thought about this and I guess we'll be fine, but it's good to discuss anyway.

> If however the predicted output is a 3d tensor (bs, f, k) (where f is some extra dimension), then the targets are now an 'integer matrix' of size (bs, f), so we'd have to find a nice and clean way to use that matrix to index the tensor.

The integers would index into the third dimension of the 3d tensor. If we rely on Theano's softmax Ops (which we should, to get the logsoftmax optimization), then we'd reshape the predictions to `(bs*f, k)` and the targets to `(bs*f,)` and everything would work out nicely.

> What if we also have the case where the predicted tensor is of shape (bs, k, f) (i.e., the softmax is now done over the 2nd axis)

Then the targets would still be `(bs, f)`. We would need to dimshuffle the predictions to `(bs, f, k)`, flatten them to `(bs*f, k)` and the targets to `(bs*f,)` and again we'd be fine.

> or if the predicted tensor was of shape (bs, k1, k2) (where the softmax is over both k1 and k2)?

That's the trickiest case. I was thinking the integer targets should be of shape `(bs,)` then, indexing into the flattened `k1*k2`. We can then flatten the predictions to `(bs, k1*k2)` and everything works fine. The alternative would be to have integer targets of `(bs, 2)` which give the position in k1 and the position in k2. Then we'd need to compute the flattened `(bs,)` version from that using the shape information from the predictions tensor. We could even support both, since the target dimensionality allows to distinguish the different cases, but I'd start with only the first option. It may be more common to use one-hot-encoded targets for more exotic cases anyway. I'm not sure how much performance differs between indexing and, err, one-hotting.

Does this make sense to you?

**Comment by christopher-beckham** [comment]
Yes that makes sense.

> I'm not sure how much performance differs between indexing and, err, one-hotting.

I'm curious about this too.

**Comment by frapa** [comment]
Any news on this?

**Comment by f0k** [comment]
> Any news on this?

Mhm, I started a spatial softmax layer implementation supporting arbitrary axes via shuffling and reshaping: https://github.com/f0k/Lasagne/commit/807763f
Then I noticed that combining this with a cross-entropy that also shuffles and reshapes breaks the optimization: https://github.com/Theano/Theano/issues/5804
This has been fixed in the meantime, but it means the shuffling/reshaping solution won't play well with Theano 0.8.2 or 0.9. I'd still include it in Lasagne, but in the next weeks I don't have any time to finish it, feel free to take over!
Another line of work is updating the Theano softmax implementation to support arbitrary axes: https://github.com/Theano/Theano/pull/5719. This is work in progress, and will be the best solution in the long run.

> Wouldn't the simplest workaround at the moment be something like the following:

Looks good, yes. (If you also want to support integer targets instead of one-hot, you will need some dimshuffling and indexing.)

> If I'm not wrong this should not have stability problems, and the gradient should be automatic. But this way the output is not the softmax, but the logsoftmax... Any way around this?

You can use `nonlinearity=None` for the Conv2DLayer and then put two separate `NonlinearityLayer` on top: One using `algo='log'` that you'd use for the loss function, and one using `algo='accurate'` that you'd use for the prediction function. You might also get away with using `algo='accurate'` for both, with a standard categorical crossentropy, this seems to have worked for at least one user (https://groups.google.com/d/msg/lasagne-users/Ltf_v6V9rTQ/tJAJrasxCQAJ).

**Comment by f0k** [comment]
Moving this to v0.3. It would be nice to have since it's commonly needed, but it's not critical for the release.

**Comment by astooke** [comment]
Have just written a version of spatial softmax for a fixed number of dimensions.  It returns not only the argmax (averaged x, y, coordinates) but also the max (thinking this will give a hint as to the sharpness of the distribution).  Haven't run it yet, so, may need to update...but just dropping this here in case it helps, or in case you see an obvious error??

```
import numpy as np
import theano
import theano.tensor as T
import lasagne.layers as L
import lasagne.init as LI
import lasagne.nonlinearities as LN

class SpatialSoftmaxLayer(L.Layer):

    def __init__(self, incoming, temperature=1, learn_temperature=False,
            name='spatial_softmax'):
        super().__init__(incoming, name=name)
        if learn_temperature:
            self.temperature = self.add_param(LI.Constant(temperature), (1,),
                name="temperature", regularizable=False)
        else:
            self.temperature = theano.shared(np.array(temperature, dtype=floatX))

    def get_output_shape_for(self, input_shape):
        assert len(input_shape) == 4
        return input_shape[0], input_shape[1], 3

    def get_output_for(self, input, **kwargs):
        """ Performs 2D spatial softmax, returns expected x & y coordinates and
        the max value out of the softmax (to give a hint how sharp it is)"""
        assert len(input.shape) == 4
        n, c, h, w = input.shape
        input_2D = T.reshape(input, (n * c, h * w))
        softmax_2D = T.softmax(input_2D / self.temperature)
        softmax_4D = T.reshape(softmax_2D, (n, c, h, w))
        pos_h = theano.shared(np.linspace(-1., 1., num=h, dtype=floatX))
        pos_w = theano.shared(np.linspace(-1., 1., num=w, dtype=floatX))
        expected_h = T.dot(pos_h, softmax_4D).sum(axis=2)
        expected_w = T.dot(softmax_4D, pos_w).sum(axis=2)
        softmax_3D = T.reshape(softmax_2D, (n, c, h * w))
        z = T.max(softmax_3D, axis=2)
        output = T.stack([expected_h, expected_w, z], axis=2)
        return output
```

**Comment by f0k** [comment]
> Have just written a version of spatial softmax for a fixed number of dimensions.

Just to clarify, what you've written is not what's usually referred to as a spatial softmax. Spatial softmax is usually understood to be a softmax across the channels per location, not vice versa. See #710.

> Haven't run it yet, so, may need to update...but just dropping this here in case it helps, or in case you see an obvious error??

Well, I think there's something wrong with the dot product and summation, and you will want to use an expression with `T.arange` instead of wrapping it as a shared variable (`np.linspace` won't like the symbolic `h` and `w` anyway). Similarly for the temperature if it's not learned; just directly set `self.temperature = temperature` instead of wrapping it as a shared variable. Other than that, it might work. If you're planning to train on this, it'd be more straightforward to just return `softmax_4D` and use categorical cross-entropy loss, though (also adapted to use the correct axes).

There's a softmax layer implementation at https://github.com/f0k/Lasagne/tree/spatial-softmax which is based on dimshuffle-reshape-softmax-reshape-dimshuffle. As a more direct implementation that precludes Theano's graph optimizations and thus has to be numerically stable itself, I've written the following:
```python
def spatial_logsoftmax(x, axis=1):
    x = x - T.max(x, axis, keepdims=True)
    return x - T.log(T.sum(T.exp(x), axis, keepdims=True))

def log_categorical_crossentropy(preds, truth, axis=-1):
    if truth.ndim == preds.ndim:
        return -T.sum(truth * preds, axis=axis)
    else:
        assert preds.ndim == 2
        assert truth.ndim == 1
        assert axis in (1, -1)
        return -preds[T.arange(truth.shape[0]), truth]
```
The former can be used as a nonlinearity (you'll need to use `functools.partial` if you want to change the `axis`, sorry about that, that's why it'd be a `Softmax` class in Lasagne), the latter as a cost function. That implementation of the cost function requires the targets to be one-hot-encoded.

**Comment by astooke** [comment]
Thanks for the tips, I have incorporated them. 
Also thanks for the clarification on spatial softmax.I'm aiming to use it in reinforcement learning, as in [https://arxiv.org/pdf/1504.00702.pdf](https://arxiv.org/pdf/1504.00702.pdf), and I did not realize there was another, more common usage of the term (!).  So, the math I'm trying to do for the expectations is like: ``expected_h = sum_{i,j}(h_{i} * z_{i,j})``,  where h is the 1-D position, z is the 2-D softmax, separate for each channel.  I checked the math on the dot-sum operations using numpy, and it seems to be correct there (both when ``h==w`` and when ``h!=w``).  Appears I may be a bit off-topic, then, but here is the newer code anyway (it builds, but I haven't tested learning with it just yet):

```python
class SpatialSoftmaxLayer(L.Layer):

    def __init__(self, incoming, temperature=1., learn_temperature=False,
            name='spatial_softmax'):
        super().__init__(incoming, name=name)
        if learn_temperature:
            self.temperature = self.add_param(LI.Constant(temperature), (1,),
                name="temperature", regularizable=False)
        else:
            self.temperature = temperature

    def get_output_shape_for(self, input_shape):
        assert len(input_shape) == 4
        return input_shape[0], input_shape[1], 3

    def get_output_for(self, input, **kwargs):
        """ Performs 2D spatial softmax, returns expected x & y coordinates and
        the max value out of the softmax (to give a hint how sharp it is)"""
        n, c, h, w = input.shape
        input_2D = T.reshape(input, (n * c, h * w))
        softmax_2D = T.nnet.softmax(input_2D / self.temperature)
        softmax_4D = T.reshape(softmax_2D, (n, c, h, w))
        # Construct ranges from -1 to 1.
        pos_h = (2 * T.arange(h, dtype=floatX) - (h - 1)) / (h - 1)
        pos_w = (2 * T.arange(w, dtype=floatX) - (w - 1)) / (w - 1)
        expected_h = T.dot(pos_h, softmax_4D).sum(axis=2)
        expected_w = T.dot(softmax_4D, pos_w).sum(axis=2)
        softmax_3D = T.reshape(softmax_2D, (n, c, h * w))
        z = T.max(softmax_3D, axis=2)
        output = T.stack([expected_h, expected_w, z], axis=2)
        return output
```
